### PR TITLE
podman-remote inspect

### PR DESCRIPTION
**Pull request by baude**
base enablement of the inspect command.

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

File: cmd/podman/varlink/io.podman.varlink [hunk 1]
```diff
@@ -1035,6 +1035,22 @@ method GenerateKubeService() -> (notimplemented: NotImplemented)
 # like that created by GenerateKube. See also [GenerateKube](GenerateKube).
 method ReplayKube() -> (notimplemented: NotImplemented)
 
+# ContainerConfig returns a container's config in string form. This call is for
```
Comment (mheon) [inline cmd/podman/varlink/io.podman.varlink]: I'm honestly OK with people using this one remotely - it seems sort of useful, even if it's a JSON string and not a literal varlink struct

File: libpod/container.go [hunk 2]
```diff
@@ -1062,3 +1067,16 @@ func networkDisabled(c *Container) (bool, error) {
...
+func (c *Container) ContainerState() (*ContainerState, error) {
+	if !c.batched {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+
+		if err := c.syncContainer(); err != nil {
+			return nil, err
+		}
+	}
+	return c.state, nil
```
Comment (mheon) [inline libpod/container.go]: Make a copy via Deepcopier first - just handing out a pointer is risky, could get changes into the database

File: libpod/container.go [hunk 3]
```diff
@@ -411,6 +411,11 @@ func (c *Container) Spec() *spec.Spec {
 	return returnSpec
 }
 
+// SpecFromState is a public wrapper for specFromState
+func (c *Container) SpecFromState() (*spec.Spec, error) {
```
Comment (mheon) [inline libpod/container.go]: I'm debating whether we want to fold this into c.Spec()... it does introduce some strangeness around not returning a consistent spec.

Comment (mheon) [reply libpod/container.go]: If we're going to keep it, rename to RunningSpec. Description: "Returns the OCI spec currently being used to run the container, or the spec the container was created with if no running spec is available"

File: cmd/podman/inspect.go [hunk 4]
```diff
@@ -102,7 +103,7 @@ func inspectCmd(c *cli.Context) error {
 }
 
 // func iterateInput iterates the images|containers the user has requested and returns the inspect data and error
-func iterateInput(ctx context.Context, c *cli.Context, args []string, runtime *libpod.Runtime, inspectType string) ([]interface{}, error) {
+func iterateInput(ctx context.Context, c *cli.Context, args []string, runtime *adapter.LocalRuntime, inspectType string) ([]interface{}, error) {
```
Comment (jwhonce) [inline cmd/podman/inspect.go]: Can we please call this adapter.GenericRuntime?

File: cmd/podman/inspect.go [hunk 5]
```diff
@@ -122,13 +123,18 @@ func iterateInput(ctx context.Context, c *cli.Context, args []string, runtime *l
 				inspectError = errors.Wrapf(err, "error getting libpod container inspect data %s", ctr.ID())
 				break
 			}
-			data, err = shared.GetCtrInspectInfo(ctr, libpodInspectData)
+			artifact, err := getArtifact(ctr)
+			if inspectError != nil {
+				inspectError = err
+				break
+			}
+			data, err = shared.GetCtrInspectInfo(ctr.Config(), libpodInspectData, artifact)
```
Comment (jwhonce) [inline cmd/podman/inspect.go]: "shared" between whom? Seems like a very questionable package name.

File: cmd/podman/inspect.go [hunk 6]
```diff
@@ -157,7 +163,12 @@ func iterateInput(ctx context.Context, c *cli.Context, args []string, runtime *l
 					inspectError = errors.Wrapf(err, "error getting libpod container inspect data %s", ctr.ID())
 					break
 				}
-				data, err = shared.GetCtrInspectInfo(ctr, libpodInspectData)
+				artifact, err := getArtifact(ctr)
+				if inspectError != nil {
```
Comment (jwhonce) [inline cmd/podman/inspect.go]: Did you mean to test "err" here in place of "inspectError"?

File: libpod/adapter/containers_remote.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+import (
+	"encoding/json"
+
+	iopodman "github.com/containers/libpod/cmd/podman/varlink"
+	"github.com/containers/libpod/libpod"
+	"github.com/containers/libpod/pkg/inspect"
+)
+
+// Inspectr returns an inspect struct from varlink
```
Comment (jwhonce) [inline libpod/adapter/containers_remote.go]: "Inspectr" spelling

File: libpod/adapter/containers_remote.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	if err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal([]byte(reply), &data); err != nil {
+		return nil, err
+	}
+	return data, err
+}
+
+func (c *Container) inspect(size bool) (*inspect.ContainerInspectData, error) {
```
Comment (jwhonce) [inline libpod/adapter/containers_remote.go]: Why duplicate the public method "Inspect()"?

File: libpod/adapter/runtime_remote.go [hunk 9]
```diff
@@ -80,21 +83,9 @@ type Container struct {
...
-	Status     string
-	//Ports            []ocicni.PortMapping
-	RootFsSize int64
-	RWSize     int64
-	Names      string
-	Labels     []map[string]string
-	//	Mounts           []string
-	// ContainerRunning bool
-	//Namespaces       []LinuxNameSpace
+	Runtime *LocalRuntime
```
Comment (jwhonce) [inline libpod/adapter/runtime_remote.go]: I would really like to call this type "GenericRuntime".

File: libpod/adapter/runtime_remote.go [hunk 10]
```diff
@@ -272,39 +263,60 @@ func (ci *ContainerImage) History(ctx context.Context) ([]*image.History, error)
...
+	// spot here.  Logging the errors for now.
+	reply, err := iopodman.ContainerConfig().Call(r.Conn, name)
+	if err != nil {
+		logrus.Error("call to container.config failed")
+	}
+	data := libpod.ContainerConfig{}
+	if err := json.Unmarshal([]byte(reply), &data); err != nil {
+		logrus.Error("failed to unmarshal container inspect data")
+	}
+	return &data
```
Comment (jwhonce) [inline libpod/adapter/runtime_remote.go]: Any chance you are trying to take the address of nil when an error occurs?

File: libpod/container.go (hunk 2 above)
Comment (baude) [reply libpod/container.go]: correction ,fixed

File: libpod/container.go (hunk 3 above)
Comment (baude) [reply libpod/container.go]: removed

File: cmd/podman/inspect.go (hunk 4 above)
Comment (baude) [reply cmd/podman/inspect.go]: sure, but lets do it outside this PR.

File: cmd/podman/inspect.go (hunk 5 above)
Comment (baude) [reply cmd/podman/inspect.go]: shared between "things" ... this has been around for a while. not part of this PR

File: cmd/podman/inspect.go (hunk 6 above)
Comment (baude) [reply cmd/podman/inspect.go]: good catch

File: libpod/adapter/containers_remote.go (hunk 7 above)
Comment (baude) [reply libpod/adapter/containers_remote.go]: fixed

File: libpod/adapter/containers_remote.go (hunk 8 above)
Comment (baude) [reply libpod/adapter/containers_remote.go]: purley by mistake I think. removed

File: libpod/adapter/runtime_remote.go (hunk 9 above)
Comment (baude) [reply libpod/adapter/runtime_remote.go]: PRs welcome!

File: libpod/adapter/runtime_remote.go (hunk 10 above)
Comment (baude) [reply libpod/adapter/runtime_remote.go]: I think the struct in data will be the default struct of ContainerConfig and we should be ok.

**Comment by baude** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
LGTM

**Comment by baude** [comment]
bot, retest this please

**Pull request merged**